### PR TITLE
[Merged by Bors] - feat: port NumberTheory.RamificationInertia

### PR DESCRIPTION
**Pull request by xroblot**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/RamificationInertia.lean [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+  IsDedekindDomain.ramificationIdx_ne_zero (ne_zero_of_mem_factors (Multiset.mem_toFinset.mp P.2))
+    (Factors.isPrime p P) (Ideal.le_of_dvd (dvd_of_mem_factors (Multiset.mem_toFinset.mp P.2)))
+#align ideal.factors.ramification_idx_ne_zero Ideal.Factors.ramificationIdx_ne_zero
+
+instance Factors.fact_ramificationIdx_neZero (P : (factors (map (algebraMap R S) p)).toFinset) :
+    NeZero (ramificationIdx (algebraMap R S) p P) :=
+  ⟨Factors.ramificationIdx_ne_zero p P⟩
+#align ideal.factors.fact_ramification_idx_ne_zero Ideal.Factors.fact_ramificationIdx_neZero
+
+set_option synthInstance.checkSynthOrder false
```
Comment (semorrison) [inline Mathlib/NumberTheory/RamificationInertia.lean:719]:
```suggestion
set_option synthInstance.checkSynthOrder false in
```

File: Mathlib/NumberTheory/RamificationInertia.lean [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+
+/-- If `P` lies over `p`, then `R / p` has a canonical map to `S / P`.
+
+This can't be an instance since the map `f : R → S` is generally not inferrable.
+-/
+def Quotient.algebraQuotientOfRamificationIdxNeZero : Algebra (R ⧸ p) (S ⧸ P) :=
+  Quotient.algebraQuotientOfLeComap (le_comap_of_ramificationIdx_ne_zero hfp.out)
+#align ideal.quotient.algebra_quotient_of_ramification_idx_ne_zero Ideal.Quotient.algebraQuotientOfRamificationIdxNeZero
+
+set_option synthInstance.checkSynthOrder false -- Porting note: this is okay by the remark below
```
Comment (semorrison) [inline Mathlib/NumberTheory/RamificationInertia.lean:478]:
```suggestion
set_option synthInstance.checkSynthOrder false in -- Porting note: this is okay by the remark below
```

File: Mathlib/NumberTheory/RamificationInertia.lean [hunk 3]
```diff
@@ -181,20 +181,20 @@ variable (f p P)
...
 extension `(S / P) : (R / p)`.
 
 We do not assume `P` lies over `p` in the definition; we return `0` instead.
 
-See `inertia_deg_algebra_map` for the common case where `f = algebra_map R S`
+See `inertiaDeg_algebraMap` for the common case where `f = algebraMap R S`
 and there is an algebra structure `R / p → S / P`.
 -/
-noncomputable def inertiaDeg [hp : p.IsMaximal] : ℕ :=
+noncomputable def inertiaDeg [_hp : p.IsMaximal] : ℕ :=
```
Comment (semorrison) [inline Mathlib/NumberTheory/RamificationInertia.lean:192]:
```suggestion
noncomputable def inertiaDeg [p.IsMaximal] : ℕ :=
```

File: Mathlib/NumberTheory/RamificationInertia.lean [hunk 4]
```diff
@@ -181,20 +181,20 @@ variable (f p P)
...
   if hPp : comap f P = p then
     @finrank (R ⧸ p) (S ⧸ P) _ _ <|
       @Algebra.toModule _ _ _ _ <|
         RingHom.toAlgebra <|
-          Ideal.Quotient.lift p (P.Quotient.mk.comp f) fun a ha =>
+          Ideal.Quotient.lift p ((Ideal.Quotient.mk P).comp f) fun _a ha =>
```
Comment (semorrison) [inline Mathlib/NumberTheory/RamificationInertia.lean:197]:
```suggestion
          Ideal.Quotient.lift p ((Ideal.Quotient.mk P).comp f) fun _ ha =>
```
And perhaps elsewhere?

File: Mathlib/NumberTheory/RamificationInertia.lean [hunk 5]
```diff
@@ -425,42 +411,39 @@ theorem finrank_quotient_map [IsDomain R] [IsDomain S] [IsDedekindDomain R] [Alg
...
     -- The nicest way to show `S ≤ span b' ⊔ pS` is by reducing both sides modulo pS.
     -- However, this would imply distinguishing between `pS` as `S`-ideal,
     -- and `pS` as `R`-submodule, since they have different (non-defeq) quotients.
     -- Instead we'll lift `x mod pS ∈ span b` to `y ∈ span b'` for some `y - x ∈ pS`.
-    intro x hx
-    have mem_span_b :
-      ((Submodule.mkQ (map (algebraMap R S) p)) x : S ⧸ map (algebraMap R S) p) ∈
-        Submodule.span (R ⧸ p) (Set.range b) :=
-      b.mem_span _
+    intro x _hx
```
Comment (semorrison) [inline Mathlib/NumberTheory/RamificationInertia.lean:432]:
```suggestion
    intro x _
```

File: Mathlib/NumberTheory/RamificationInertia.lean [hunk 6]
```diff
@@ -496,48 +475,47 @@ def Quotient.algebraQuotientOfRamificationIdxNeZero : Algebra (R ⧸ p) (S ⧸ P
...
 def powQuotSuccInclusion (i : ℕ) :
-    Ideal.map (P ^ e).Quotient.mk (P ^ (i + 1)) →ₗ[R ⧸ p] Ideal.map (P ^ e).Quotient.mk (P ^ i)
-    where
+    Ideal.map (Ideal.Quotient.mk (P ^ e)) (P ^ (i + 1)) →ₗ[R ⧸ p]
+    Ideal.map (Ideal.Quotient.mk (P ^ e)) (P ^ i) where
   toFun x := ⟨x, Ideal.map_mono (Ideal.pow_le_pow i.le_succ) x.2⟩
-  map_add' x y := rfl
-  map_smul' c x := rfl
+  map_add' _x _y := rfl
+  map_smul' _c _x := rfl
```
Comment (semorrison) [inline Mathlib/NumberTheory/RamificationInertia.lean:494]:
```suggestion
  map_add' _ _ := rfl
  map_smul' _ _ := rfl
```

**Comment by semorrison** [comment]
bors d+

File: Mathlib/NumberTheory/RamificationInertia.lean (hunk 1 above)
Comment (xroblot) [reply Mathlib/NumberTheory/RamificationInertia.lean:719]: This does seem to be doing what we think it is doing: adding it make the instance invisible later in the file... For example:
```lean
set_option synthInstance.checkSynthOrder false in -- Porting note: this is okay by the remark below
-- In this file, the value for `f` can be inferred.
attribute [local instance] Ideal.Quotient.algebraQuotientOfRamificationIdxNeZero

@[simp]
theorem Quotient.algebraMap_quotient_of_ramificationIdx_neZero (x : R) :
    algebraMap (R ⧸ p) (S ⧸ P) (Ideal.Quotient.mk p x) = Ideal.Quotient.mk _ (f x) := rfl
```
The `Algebra` structure cannot be synthesize in the theorem. 

I'll leave it the way it is.

**Pull request closed without merging**